File: momentra/momentra/Shell/BusinessActive/BusinessActionRegistry.swift
import Foundation
import SwiftUI

/// V019 Business Quick Add capability codes (S4 + Ops). Fail-closed like Personal.
enum BusinessActionCode: String, CaseIterable, Equatable {
    case expenseCreate = "EXPENSE_CREATE"
    case revenueRecord = "REVENUE_RECORD"
    case invoiceCreate = "INVOICE_CREATE"
    case memberManage = "MEMBER_MANAGE"
    case vendorManage = "VENDOR_MANAGE"
    case issueCreate = "ISSUE_CREATE"
    case slaManage = "SLA_MANAGE"
}

enum BusinessActionDestination: Equatable {
    case spend
    case vendor
    case issue
    case sla
    case revenue
    case invoice
    case members
}

struct BusinessActionTile: Identifiable {
    var id: String { "\(code.rawValue)-\(label)" }
    let code: BusinessActionCode
    let label: String
    let icon: String
    let colors: [Color]
    let destination: BusinessActionDestination
    let enabledWhenMomentActive: Bool
}

enum BusinessActionRegistry {
    static func destination(for capabilityCode: String) -> BusinessActionDestination? {
        switch capabilityCode.uppercased() {
        case BusinessActionCode.expenseCreate.rawValue: return .spend
        case BusinessActionCode.vendorManage.rawValue: return .vendor
        case BusinessActionCode.issueCreate.rawValue: return .issue
        case BusinessActionCode.slaManage.rawValue: return .sla
        case BusinessActionCode.revenueRecord.rawValue: return .revenue
        case BusinessActionCode.invoiceCreate.rawValue: return .invoice
        case BusinessActionCode.memberManage.rawValue: return .members
        default: return nil
        }
    }

    static func defaultCodes() -> [BusinessActionCode] {
        [.expenseCreate, .revenueRecord, .invoiceCreate, .memberManage, .vendorManage, .issueCreate, .slaManage]
    }

    /// Empty / nil capabilities fail closed (mirror Personal).
    static func isDestinationEnabled(_ capabilities: [String]?, destination target: BusinessActionDestination) -> Bool {
        guard let capabilities, !capabilities.isEmpty else {
            return false
        }
        return capabilities.contains { destination(for: $0) == target }
    }

    static func destination(for kind: BusinessQuickAddKind) -> BusinessActionDestination? {
        switch kind {
        case .expense, .spendEntry, .requestApproval, .budgetReview: return .spend
        case .updateVendor: return .vendor
        case .reportIssue, .logImprovement: return .issue
        case .slaCheck: return .sla
        case .revenue: return .revenue
        case .invoice: return .invoice
        default: return nil
        }
    }

    /// Unmapped kinds (update/memory) stay available when moment is active.
    static func isKindEnabled(_ kind: BusinessQuickAddKind, capabilities: [String]?) -> Bool {
        guard let dest = destination(for: kind) else { return true }
        return isDestinationEnabled(capabilities, destination: dest)
    }

    /// Builds hub tiles.
    /// - `nil` capabilityCodes → default V019 codes (catalog / tests)
    /// - empty array → fail closed (no tiles)
    /// - non-empty → destination-level enablement
    static func tiles(
        hasActiveMoment: Bool,
        capabilityCodes: [String]? = nil
    ) -> [BusinessActionTile] {
        let effectiveCaps = capabilityCodes ?? defaultCodes().map(\.rawValue)
        return catalogTiles().compactMap { tile in
            if !isDestinationEnabled(effectiveCaps, destination: tile.destination) {
                return nil
            }
            return BusinessActionTile(
                code: tile.code,
                label: tile.label,
                icon: tile.icon,
                colors: tile.colors,
                destination: tile.destination,
                enabledWhenMomentActive: tile.enabledWhenMomentActive && hasActiveMoment
            )
        }
    }

    private static func catalogTiles() -> [BusinessActionTile] {
        let accent = Color(red: 0.506, green: 0.549, blue: 0.973)
        let soft = Color(red: 0.647, green: 0.706, blue: 0.988)
        return [
            BusinessActionTile(
                code: .expenseCreate,
                label: "Expense",
                icon: "creditcard",
                colors: [accent, soft],
                destination: .spend,
                enabledWhenMomentActive: true
            ),
            BusinessActionTile(
                code: .revenueRecord,
                label: "Revenue",
                icon: "chart.line.uptrend.xyaxis",
                colors: [Color(hex: "#14B8A6"), Color(hex: "#2DD4BF")],
                destination: .revenue,
                enabledWhenMomentActive: true
            ),
            BusinessActionTile(
                code: .invoiceCreate,
                label: "Invoice",
                icon: "doc.text",
                colors: [Color(hex: "#3B82F6"), Color(hex: "#60A5FA")],
                destination: .invoice,
                enabledWhenMomentActive: true
            ),
            BusinessActionTile(
                code: .memberManage,
                label: "People",
                icon: "person.2",
                colors: [Color(hex: "#6366F1"), Color(hex: "#818CF8")],
                destination: .members,
                enabledWhenMomentActive: true
            ),
        ]
    }
}

File: apk/app/src/main/java/com/example/momentra/ui/shell/business/BusinessActionRegistry.kt
package com.example.momentra.ui.shell.business

/**
 * Mapper from V019 Business capability / action codes to Quick Add destinations.
 * Empty capabilities fail closed until bootstrap fills V019 codes (parity with Personal).
 */
object BusinessActionRegistry {

    const val EXPENSE_CREATE = "EXPENSE_CREATE"
    const val REVENUE_RECORD = "REVENUE_RECORD"
    const val INVOICE_CREATE = "INVOICE_CREATE"
    const val MEMBER_MANAGE = "MEMBER_MANAGE"
    const val VENDOR_MANAGE = "VENDOR_MANAGE"
    const val ISSUE_CREATE = "ISSUE_CREATE"
    const val SLA_MANAGE = "SLA_MANAGE"

    enum class Destination {
        SPEND,
        VENDOR,
        ISSUE,
        SLA,
        REVENUE,
        INVOICE,
        MEMBERS,
    }

    fun destinationFor(capabilityCode: String): Destination? = when (capabilityCode.uppercase()) {
        EXPENSE_CREATE -> Destination.SPEND
        VENDOR_MANAGE -> Destination.VENDOR
        ISSUE_CREATE -> Destination.ISSUE
        SLA_MANAGE -> Destination.SLA
        REVENUE_RECORD -> Destination.REVENUE
        INVOICE_CREATE -> Destination.INVOICE
        MEMBER_MANAGE -> Destination.MEMBERS
        else -> null
    }

    /** Revenue/Invoice are V019-mapped to BUSINESS_RUNWAY only. */
    fun isRunwayFinanceEnabled(momentTypeCode: String?): Boolean =
        momentTypeCode?.uppercase() == "BUSINESS_RUNWAY"

    fun isDestinationEnabled(capabilities: List<String>, destination: Destination): Boolean {
        if (capabilities.isEmpty()) return false
        return capabilities.any { destinationFor(it) == destination }
    }

    fun enabledDestinations(capabilities: List<String>): Set<Destination> {
        if (capabilities.isEmpty()) return emptySet()
        return capabilities.mapNotNull { destinationFor(it) }.toSet()
    }
}

fun BusinessQuickAddKind.registryDestination(): BusinessActionRegistry.Destination? = when (this) {
    BusinessQuickAddKind.EXPENSE,
    BusinessQuickAddKind.SPEND_ENTRY,
    BusinessQuickAddKind.REQUEST_APPROVAL,
    BusinessQuickAddKind.BUDGET_REVIEW,
    -> BusinessActionRegistry.Destination.SPEND
    BusinessQuickAddKind.UPDATE_VENDOR -> BusinessActionRegistry.Destination.VENDOR
    BusinessQuickAddKind.REPORT_ISSUE,
    BusinessQuickAddKind.LOG_IMPROVEMENT,
    -> BusinessActionRegistry.Destination.ISSUE
    BusinessQuickAddKind.SLA_CHECK -> BusinessActionRegistry.Destination.SLA
    BusinessQuickAddKind.REVENUE -> BusinessActionRegistry.Destination.REVENUE
    BusinessQuickAddKind.INVOICE -> BusinessActionRegistry.Destination.INVOICE
    else -> null
}

/** Unmapped kinds (update/memory/chrome) stay available when moment is active. */
fun BusinessQuickAddKind.isCapabilityEnabled(capabilities: List<String>): Boolean {
    val dest = registryDestination() ?: return true
    return BusinessActionRegistry.isDestinationEnabled(capabilities, dest)
}
